src/workerd/api/streams/standard.c++

```diff
@@ -461,25 +461,36 @@ jsg::Promise<void> maybeRunAlgorithm(
   // that it is properly mapped to a jsg::Promise, but if the Promise
   // throws synchronously, we have to convert that synchronous throw
   // into a proper rejected jsg::Promise.
-
-  if (IoContext::hasCurrent()) {
-    auto& ioContext = IoContext::current();
-    KJ_IF_MAYBE(algorithm, maybeAlgorithm) {
-      return js.tryCatch([&] {
-        return (*algorithm)(js, kj::fwd<decltype(args)>(args)...);
-      }, [&](jsg::Value&& exception) {
-        return js.rejectedPromise<void>(kj::mv(exception));
-      }).then(js, ioContext.addFunctor(kj::mv(onSuccess)),
-                  ioContext.addFunctor(kj::mv(onFailure)));
-    }
-  } else {
-    KJ_IF_MAYBE(algorithm, maybeAlgorithm) {
-      return js.tryCatch([&] {
-        return (*algorithm)(js, kj::fwd<decltype(args)>(args)...);
-      }, [&](jsg::Value&& exception) {
-        return js.rejectedPromise<void>(kj::mv(exception));
-      }).then(js, kj::mv(onSuccess), kj::mv(onFailure));
-    }
+  KJ_IF_SOME(algorithm, maybeAlgorithm) {
+    // We need two layers of tryCatch here, unfortunately. The inner layer
+    // covers the algorithm implementation itself and is our typical error
+    // handling path. It ensures that if the algorithm throws an exception,
+    // that is properly converted in to a rejected promise that is *then*
+    // handled by the onFailure handler that is passed in. The outer tryCatch
+    // handles the rare and generally unexpected failure of the calls to
+    // .then() itself, which can throw JS exceptions synchronously in certain
+    // rare cases. For those we return a rejected promise but do not call the
+    // onFailure case since such errors are generally indicative of a fatal
+    // condition in the isolate (e.g. out of memory, other fatal exception, etc).
+    return js.tryCatch([&] {
```

Comment:
If it's not safe to throw `JsExceptionThrown` from here, then `js.tryCatch()` is not sufficient to fix that, because it does not catch termination exceptions.